### PR TITLE
[SPARK-35069][SQL] TRANSFORM forbidden  `DISTICNT` and `ALL`, also make the error clear

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
According to https://github.com/apache/spark/pull/29087#discussion_r612267050,  add UT in `transform.sql`

It seems that distinct is not recognized as a reserved word here

```
-- !query
explain extended SELECT TRANSFORM(distinct b, a, c)
                   USING 'cat' AS (a, b, c)
                 FROM script_trans
                 WHERE a <= 4
-- !query schema
struct<plan:string>
-- !query output
== Parsed Logical Plan ==
'ScriptTransformation [*], cat, [a#x, b#x, c#x], ScriptInputOutputSchema(List(),List(),None,None,List(),List(),None,None,false)
+- 'Project ['distinct AS b#x, 'a, 'c]
   +- 'Filter ('a <= 4)
      +- 'UnresolvedRelation [script_trans], [], false

== Analyzed Logical Plan ==
org.apache.spark.sql.AnalysisException: cannot resolve 'distinct' given input columns: [script_trans.a, script_trans.b, script_trans.c]; line 1 pos 34;
'ScriptTransformation [*], cat, [a#x, b#x, c#x], ScriptInputOutputSchema(List(),List(),None,None,List(),List(),None,None,false)
+- 'Project ['distinct AS b#x, a#x, c#x]
   +- Filter (a#x <= 4)
      +- SubqueryAlias script_trans
         +- View (`script_trans`, [a#x,b#x,c#x])
            +- Project [cast(a#x as int) AS a#x, cast(b#x as int) AS b#x, cast(c#x as int) AS c#x]
               +- Project [a#x, b#x, c#x]
                  +- SubqueryAlias script_trans
                     +- LocalRelation [a#x, b#x, c#x]
```

Hive's error 
image

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added Ut

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
FYI @cloud-fan

File: sql/core/src/test/resources/sql-tests/results/transform.sql.out [hunk 1]
```diff
@@ -638,3 +639,15 @@ struct<b:string>
 1
 2
 3
+
+
+-- !query
+SELECT TRANSFORM(distinct b, a, c)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/transform.sql.out:645]: ah the parser treats it as `distinct as b`.

Can we improve the parser to recognize `distinct` correctly and give a clear error message?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:645]: BTW why do we support `TRANSFORM(a as b)`? is it from Hive? Normal function parameters can't have alias.

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:645]: > BTW why do we support `TRANSFORM(a as b)`? is it from Hive? Normal function parameters can't have alias.

Hive not support this. 

image

may be we should not use 
```
namedExpressionSeq
    : namedExpression (',' namedExpression)*
    ;

(SELECT kind=TRANSFORM '(' namedExpressionSeq ')'
            | kind=MAP namedExpressionSeq
            | kind=REDUCE namedExpressionSeq)
```

maybe 
```
expressionSeq 
    : expression(',' expression) 
    ;

(SELECT kind=TRANSFORM '(' expressionSeq ')'
            | kind=MAP expressionSeq
            | kind=REDUCE eExpressionSeq)
```

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:645]: yea, let's fix it.

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:645]: > yea, let's fix it as well.

Should I fix both issue in this pr or  in separated PR?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:645]: let's have 2 PRs. forbid distinct and forbid alias are two things.

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:645]: > let's have 2 PRs. forbid distinct and forbid alias are two things.

I have make a new ticket for this one. and changed the title.

How about current

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala [hunk 2]
```diff
@@ -88,6 +88,11 @@ object QueryParsingErrors {
     new ParseException("DISTRIBUTE BY is not supported", ctx)
   }
 
+  def transformNotSupportQuantifierError(ctx: ParserRuleContext): Throwable = {
+    new ParseException("TRANSFORM not support quantifier DISTINCT/ALL " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:92]: not support -> does not support

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala [hunk 3]
```diff
@@ -88,6 +88,11 @@ object QueryParsingErrors {
...
+      "for input expression seq", ctx)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:93]: `for input expression seq` -> `in inputs`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:92]: `quantifier DISTINCT/ALL` -> `DISTINCT/ALL`

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:92]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:93]: Done

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**